tsfile/src/main/java/org/apache/iotdb/tsfile/read/query/executor/TsFileExecutor.java

```diff
@@ -54,6 +58,25 @@ public TsFileExecutor(IMetadataQuerier metadataQuerier, IChunkLoader chunkLoader
 
   @Override
   public QueryDataSet execute(QueryExpression queryExpression) throws IOException {
+    // bloom filter
+    BloomFilter bloomFilter = metadataQuerier.getWholeFileMetadata().getBloomFilter();
+    List<Path> filteredSeriesPath = new ArrayList<>();
+    if (bloomFilter != null) {
+      for (Path path : queryExpression.getSelectedSeries()) {
+        if (bloomFilter.contains(path.getFullPath())) {
+          filteredSeriesPath.add(path);
+        }
+      }
+      if (filteredSeriesPath.isEmpty()) {
+        // return an empty QueryDataSet
+        LOG.debug(
+            "This tsfile not contains all paths which specified in selected series, return empty dataset");
+        return new DataSetWithoutTimeGenerator(Collections.emptyList(), Collections.emptyList(),
+            Collections.emptyList());
+      }
+    }
+    queryExpression.setSelectSeries(filteredSeriesPath);
```

Comment:
actually, I wonder why the performance is improved.
After you do `queryExpression.setSelectSeries(filteredSeriesPath);`,  you have to do `metadataQuerier.loadChunkMetaDatas(queryExpression.getSelectedSeries());` right?

Reply:
and, when you run experiments, you'd better discuss in several cases:
if all paths exist; 
if most of paths  exist;
if a few paths exist;
if no paths exist.

Reply:
1. If the file contains no required timeseries, it will be very fast using the bloom filter.
2. Even if only some of the timeseries are filtered, it reduces the lower calls to search and compare String. As you know, String comparison can be quite costly.

Reply:
For this part which is only be used by readOnlyTsfile, if one of the paths not in tsfile, we should just return an empty dataset or throw an exception. Because the original behavior is this.

Reply:
According to the code, I think the bloom filter takes effect in the iotdb query process, not in the reading process of TsFile. As you know, `metadataQuerier.loadChunkMetaDatas(queryExpression.getSelectedSeries())` is used only in TsFile.

Reply:
The bloom filter will play a vital role in the following example scenario:

Assume:
- The measurement scheme registered under the storage group `root.sg` is: s0,s1,s2,...,s3200
- The TsFile file1 only contains data from `root.sg.d0.s0`

Then query data from series paths: `root.sg.d0.s1`,`root.sg.d0.s2`,...,`root.sg.d0.s3200`.

As shown in `DeviceMetaDataCache.get`, without the bloom filter, the TsDeviceMetaData of `root.sg.d0` will be read from disk 3200 times. The existent metadata cache mechanism won't help here because List<ChunkMetaData> of `root.sg.d0.s0` doesn't exist in file1.

Now with the bloom filter in the TsFileMetaData, some of the 3200 queried series can be judged to be nonexistent in the file1, therefore eliminating the need to read TsDeviceMetaData from disk.

Reply:
> The bloom filter will play a vital role in the following example scenario:
> 
> Assume:
> 
> * The measurement scheme registered under the storage group `root.sg` is: s0,s1,s2,...,s3200
> * The TsFile file1 only contains data from `root.sg.d0.s0`
> 
> Then query data from series paths: `root.sg.d0.s1`,`root.sg.d0.s2`,...,`root.sg.d0.s3200`.
> 
> As shown in `DeviceMetaDataCache.get`, without the bloom filter, the TsDeviceMetaData of `root.sg.d0` in file1 will be read from disk 3200 times. The existent metadata cache mechanism won't help here because it will only caches List <ChunkMetaData> of those existent series while `root.sg.d0.s0` doesn't exist in file1.
> 
> Now with the bloom filter in the TsFileMetaData, some of the 3200 queried series can be judged to be nonexistent in the file1, therefore eliminating the next step of reading TsDeviceMetaData from disk.

well, but how about if all 3200 paths exist in the file? is there any improvement?

> the TsDeviceMetaData of root.sg.d0 in file1 will be read from disk 3200 times.

By the way, I think it should be improvement...

Reply:
> well, but how about if all 3200 paths exist in the file? is there any improvement?

If all 3200 paths have their data existent in this tsfile, that mean this tsfile will also contain their ChunkMetaData, therefore when this tsfile is first read for a certain path, the certain path's and other "hot" sensors' ChunkMetaData will be cached (you can compare it to "disk prefetch").

Reply:
Ok, got it. 
But I think we also need to claim it clearly the improvement happens at which case. @SilverNarcissus